bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/grpc/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -0,0 +1,216 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.grpc;
+
+import io.grpc.Call;
+import io.grpc.Channel;
+import io.grpc.ClientInterceptor;
+import io.grpc.ClientInterceptors.CheckedForwardingCall;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+
+import java.io.IOException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import com.google.api.client.util.Clock;
+import com.google.api.client.util.Preconditions;
+import com.google.auth.oauth2.AccessToken;
+import com.google.auth.oauth2.OAuth2Credentials;
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Client interceptor that authenticates all calls by binding header data provided by a credential.
+ * Typically this will populate the Authorization header but other headers may also be filled out.
+ *
+ * <p> Uses the new and simplified Google auth library:
+ * https://github.com/google/google-auth-library-java</p>
+ *
+ * <p> TODO: COPIED FROM io.grpc.auth.ClientAuthInterceptor.  The logic added here for
+ * initialization and locking could be moved back to gRPC.  This implementation takes advantage
+ * of the fact that all of the Bigtable endpoints are OAuth2 based.  It uses the OAuth AccessToken
+ * to get the token value and next refresh time.  The refresh is scheduled asynchronously.</p>
+ */
+public class RefreshingOAuth2CredentialsInterceptor implements ClientInterceptor {
+
+  private enum CacheState {
+    Good,
+    Stale,
+    Expired
+  }
+
+  private static class TokenCache {
+    final IOException exception;
+    final String header;
+    final long staleTime;
+    final long expiresTime;
+
+    public TokenCache(String cachedHeader, long staleTime, long expiresTime) {
+      super();
+      this.exception = null;
+      this.header = cachedHeader;
+      this.staleTime = staleTime;
+      this.expiresTime = expiresTime;
+    }
+
+    public TokenCache(IOException exception) {
+      super();
+      this.exception = exception;
+      this.header = null;
+      this.staleTime = -1;
+      this.expiresTime = -1;
+    }
+
+    public CacheState getCacheState() {
+      long now = System.currentTimeMillis();
+      if (now > expiresTime) {
+        return CacheState.Expired;
+      } else if (now > staleTime) {
+        return CacheState.Stale;
+      } else {
+        return CacheState.Good;
+      }
+    }
+  }
+  
+  /**
+   * This specifies how far in advance of a token expiration do we consider the token stale.
+   */
+  private static final int TOKEN_STALENESS_MS = 60 * 1000;
+  private static final int TOKEN_EXPIRES_MS = 1 * 1000;
```

Comment:
Explain what TOKEN_EXPIRES_MS? is it how far in advance of token expiration we consider the token expired?